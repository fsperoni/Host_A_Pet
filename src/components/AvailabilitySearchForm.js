import React, { useState, useContext } from "react";
import Alert from "./Alert";
import HostAPetApi from "./Api";
import UserContext from '../hooks/useUserContext';

/** Search Availability form. */

const AvailabilitySearchForm = ({ avails, setAvails, roles}) => {
  const { currentUser } = useContext(UserContext);
  const min = new Date().toISOString().substring(0,10);
  const [formData, setFormData] = useState({
    startDate: min,
    endDate: min,
    roleId: ""
  });
  const [formErrors, setFormErrors] = useState([]);
  

  /** Handle form submit */
  const handleSubmit = async (evt) => {
    evt.preventDefault();
    try {
      const result = await HostAPetApi.getAllAvailabilities(currentUser.username, formData);
      setAvails(result);
      setFormData({
        startDate: min,
        endDate: min,
        roleId: ""
      });
      setFormErrors([]);
    } catch (err) {
      setAvails([]);
      setFormErrors(err);
    }
  }

  /** Handle form field changes */
  const handleChange = (evt) => {
    const { name, value } = evt.target;
    setFormData(data => ({ ...data, [name]: value }));
    setFormErrors([]);
  }

  /** Handle form cancel button */
  const handleCancel = (evt) => {
    evt.preventDefault();
    setFormData({
      startDate: min,
      endDate: min,
      roleId: ""
    });
    setFormErrors([]);
  }

  const generateOptions = () => {
    return roles.map(role => (
      <option key={role.id} value={role.id}>{role.name}</option>
    ))
  }
  const options = generateOptions();
  return (
    <div className="AvailabilityAddForm mt-3">
      <div className="container">
        <h3 className="mb-3">Start your search!</h3>
        <div className="card">
          <div className="card-body">
            <form onSubmit={handleSubmit}>
              <div className="form-group">
                <label>Looking for a:</label>
                <select name="roleId" className="form-select" onChange={handleChange}
                  value={formData.roleId}>
                <option value="">Select one</option>
                  {options}
                </select>
              </div>
              <div className="form-group">
                <label>From:</label>
                <input name="startDate" className="form-control" onChange={handleChange}
                  value={formData.startDate} type="date" min={min}
                />
              </div>
              <div className="form-group">
                <label>Until:</label>
                <input name="endDate" className="form-control" onChange={handleChange}
                  value={formData.endDate} type="date" min={min}
                />
              </div>
              {formErrors.length>0 && <Alert type="danger" messages={formErrors} />}
              <div className="container mt-4">
              <button 
                className="btn btn-primary btn-block"
                onClick={handleSubmit}
              > 
                Search
              </button>
              <button
                className="btn btn-warning btn-block ms-4"
                onClick={handleCancel}
              >
                Cancel
              </button>
            </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  );
}

export default AvailabilitySearchForm;