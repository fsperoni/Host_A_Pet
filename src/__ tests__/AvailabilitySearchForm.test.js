import { render } from "@testing-library/react";
import AvailabilitySearchForm from "../components/AvailabilitySearchForm";
import UserContext from "../hooks/useUserContext";

const {
  currentAdminUser,
  roles,
} = require("./_testCommon");

// smoke test
it('renders without crashing', function() {
  render(
    <UserContext.Provider
          value={{ currentUser: currentAdminUser() , function() {}}}>
      <AvailabilitySearchForm setDateRange={function() {}} setAvails={function() {}}
        setBookSuccess={function() {}} setNotFound={function() {}} roles={roles()} setBookError={function() {}}/>
    </UserContext.Provider>);
});

//snapshot test
it("matches snapshot", function() {
  const {asFragment} = render(
    <UserContext.Provider
          value={{ currentUser: currentAdminUser() , function() {}}}>
      <AvailabilitySearchForm setDateRange={function() {}} setAvails={function() {}}
        setBookSuccess={function() {}} setNotFound={function() {}} roles={roles()} setBookError={function() {}}/>
    </UserContext.Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("should have 'Start your search!' in the component.", function() {
  const { getByText } = render(
    <UserContext.Provider
          value={{ currentUser: currentAdminUser() , function() {}}}>
      <AvailabilitySearchForm setDateRange={function() {}} setAvails={function() {}}
        setBookSuccess={function() {}} setNotFound={function() {}} roles={roles()} setBookError={function() {}}/>
    </UserContext.Provider>
  );
  expect(getByText("Start your search!")).toBeInTheDocument();
});