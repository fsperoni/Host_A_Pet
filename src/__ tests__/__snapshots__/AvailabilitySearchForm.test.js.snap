// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="AvailabilityAddForm mt-3"
  >
    <div
      class="container"
    >
      <h3
        class="mb-3"
      >
        Start your search!
      </h3>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <form>
            <div
              class="form-group"
            >
              <label>
                Looking for a:
              </label>
              <select
                class="form-select"
                name="roleId"
              >
                <option
                  value=""
                >
                  Select one
                </option>
                <option
                  value="1"
                >
                  Host
                </option>
                <option
                  value="2"
                >
                  Pet Owner
                </option>
              </select>
            </div>
            <div
              class="form-group"
            >
              <label>
                From:
              </label>
              <input
                class="form-control"
                min="2022-04-02"
                name="startDate"
                type="date"
                value="2022-04-02"
              />
            </div>
            <div
              class="form-group"
            >
              <label>
                Until:
              </label>
              <input
                class="form-control"
                min="2022-04-02"
                name="endDate"
                type="date"
                value="2022-04-02"
              />
            </div>
            <div
              class="container mt-4"
            >
              <button
                class="btn btn-primary btn-block"
              >
                Search
              </button>
              <button
                class="btn btn-warning btn-block ms-4"
              >
                Cancel
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
